File: gto-support-util/src/main/java/org/ccci/gto/android/common/util/LocaleUtils.java
package org.ccci.gto.android.common.util;

import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.Locale;
import java.util.Map;

public class LocaleUtils {
    // define a few fixed fallbacks
    static final Map<String, String> FALLBACKS = new HashMap<>();
    static {
        // macrolanguage fallbacks from https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry
        // Malagasy macrolanguage
        FALLBACKS.put("bhr", "mg");
        FALLBACKS.put("bjq", "mg");
        FALLBACKS.put("bmm", "mg");
        FALLBACKS.put("bzc", "mg");
        FALLBACKS.put("msh", "mg");
        FALLBACKS.put("plt", "mg");
        FALLBACKS.put("skg", "mg");
        FALLBACKS.put("tdx", "mg");
        FALLBACKS.put("tkg", "mg");
        FALLBACKS.put("txy", "mg");
        FALLBACKS.put("xmv", "mg");
        FALLBACKS.put("xmw", "mg");

        // Malay macrolanguage
        FALLBACKS.put("mfa", "ms");
        FALLBACKS.put("pse", "ms");
        FALLBACKS.put("zlm", "ms");
    }

    private static final LocaleCompat COMPAT;
    static {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            COMPAT = new LocaleCompat.Base();
        } else {
            COMPAT = new LocaleCompat.Nougat();
        }
    }

    // region Language fallback methods

    public static void addFallback(@NonNull final String locale, @NonNull final String fallback) {
        if (FALLBACKS.get(locale) != null) {
            throw new IllegalStateException(locale + " already has a fallback language defined");
        }
        FALLBACKS.put(locale, fallback);
    }

    @Nullable
    public static Locale getFallback(@NonNull final Locale locale) {
        return COMPAT.getFallback(locale);
    }

    @NonNull
    public static Locale[] getFallbacks(@NonNull final Locale locale) {
        return COMPAT.getFallbacks(locale);
    }

    @NonNull
    public static Locale[] getFallbacks(@NonNull final Locale... locales) {
        final LinkedHashSet<Locale> outputs = new LinkedHashSet<>();

        // generate fallbacks for all provided locales
        for (final Locale locale : locales) {
            Collections.addAll(outputs, getFallbacks(locale));
        }

        return outputs.toArray(new Locale[0]);
    }
    // endregion Language fallback methods
}

File: gto-support-util/src/main/kotlin/org/ccci/gto/android/common/util/LocaleUtils.kt
package org.ccci.gto.android.common.util

import android.icu.util.ULocale
import android.os.Build
import androidx.annotation.RequiresApi
import java.util.Locale

object LocaleUtils {
    private val COMPAT = when {
        Build.VERSION.SDK_INT < Build.VERSION_CODES.N -> Compat.Base()
        else -> Compat.Nougat()
    }

    private val FIXED_FALLBACKS = mutableMapOf(
        // macro language fallbacks https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry
        // Malagasy macro language
        "bhr" to "mg",
        "bjq" to "mg",
        "bmm" to "mg",
        "bzc" to "mg",
        "msh" to "mg",
        "plt" to "mg",
        "skg" to "mg",
        "tdx" to "mg",
        "tkg" to "mg",
        "txy" to "mg",
        "xmv" to "mg",
        "xmw" to "mg",

        // Malay macrolanguage
        "mfa" to "ms",
        "pse" to "ms",
        "zlm" to "ms"
    )

    // region Language fallback methods
    @JvmStatic
    fun addFallback(locale: String, fallback: String) {
        require(FIXED_FALLBACKS[locale] == null) { "$locale already has a fallback language defined" }
        FIXED_FALLBACKS[locale] = fallback
    }

    internal fun generateFallbacksSequence(locale: Locale) = COMPAT.generateFallbacksSequence(locale)
    // endregion Language fallback methods

    private sealed interface Compat {
        fun generateFallbacksSequence(locale: Locale): Sequence<Locale>

        open class Base : Compat {
            override fun generateFallbacksSequence(locale: Locale): Sequence<Locale> {
                val builder = Locale.Builder().setLocaleSafe(locale)
                return generateSequence {
                    val currentLocale = builder.buildOrNull() ?: return@generateSequence null
                    val directFallback = FIXED_FALLBACKS[currentLocale.toLanguageTag()]
                    when {
                        directFallback != null -> {
                            val fallbackLocale = Locale.forLanguageTag(directFallback)
                            builder.setLocaleSafe(fallbackLocale)
                            fallbackLocale
                        }
                        currentLocale.extensionKeys.isNotEmpty() -> builder.clearExtensions().build()
                        currentLocale.variant.isNotEmpty() -> builder.setVariant(null).build()
                        currentLocale.country.isNotEmpty() -> builder.setRegion(null).build()
                        currentLocale.script.isNotEmpty() -> builder.setScript(null).build()
                        else -> {
                            builder.clear()
                            null
                        }
                    }
                }.distinct()
            }
        }

        @RequiresApi(Build.VERSION_CODES.N)
        class Nougat : Base() {
            override fun generateFallbacksSequence(locale: Locale) =
                generateSequence(ULocale.forLocale(locale)) {
                    val fixed = FIXED_FALLBACKS[it.toLanguageTag()]
                    when {
                        // fixed fallback
                        fixed != null -> ULocale.forLanguageTag(fixed)
                        // remove extensions as the fallback if any are defined
                        it.extensionKeys.isNotEmpty() -> ULocale.Builder().setLocale(it).clearExtensions().build()
                        // use normal fallback behavior
                        else -> it.fallback.takeUnless { it == ULocale.ROOT }
                    }
                }.drop(1).map { it.toLocale() }.distinct()
        }
    }
}
